File: solutions/lua/scrabble-score/1/scrabble-score.lua
local xs = {
	["AEIOULNRST"] = 1,
	["DG"] = 2,
	["BCMP"] = 3,
	["FHVWY"] = 4,
	["K"] = 5,
	["JX"] = 8,
	["QZ"] = 10
}

local function score(word)
	local res = 0
	if word == nil then
		return res
	end
	for c in string.gmatch(word, ".") do
		for k, v in pairs(xs) do
			if string.match(string.upper(c), "[" .. k .. "]") then
				res = res + v
				break
			end
		end
	end
	return res
end

return { score = score }

File: solutions/ocaml/scrabble-score/1/scrabble_score.ml
open Base

let xs : (string * int) list = [
  ("AEIOULNRST", 1);
  ("DG", 2);
  ("BCMP", 3);
  ("FHVWY", 4);
  ("K", 5);
  ("JX", 8);
  ("QZ", 10);
]

let score (word: string): int =
  if String.is_empty word then 0 else
  List.sum (module Int) (String.to_list word) ~f:(fun c ->
    List.sum (module Int) xs ~f:(fun (k, v) ->
      let matches = List.exists ~f:(fun c2 -> Char.(c2 = uppercase c)) (String.to_list k) in
      v * Bool.to_int matches
    )
  )
